src/Data/Bin/Ordering.agda

```diff
@@ -0,0 +1,790 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Ordering on Bin and certain its properties.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Bin.Ordering where
+
+open import Data.Bin.Base
+open import Data.Bin.Properties
+open import Data.Nat as ℕ using (ℕ; z≤n; s≤s)
+open import Data.Nat.Properties as ℕₚ using (m+n∸m≡n; m+n∸n≡m)
+open import Data.Product using (_,_; proj₁; proj₂)
+open import Data.Sum
+open import Function
+open import Level using () renaming (zero to 0ℓ)
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality as P
+open import Relation.Binary.PropositionalEquality.Core using (≢-sym)
+import Relation.Binary.Reasoning.Base.Triple as InequalityReasoning
+open import Relation.Nullary using (¬_; yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+
+open Bin
+
+-- The ordering relation
+
+infix 4 _<_ _>_ _≤_
+
+data _<_ : Rel Bin 0ℓ  where
+  0<even    : ∀ {x} → zero < 2[1+ x ]
+  0<odd     : ∀ {x} → zero < 1+[2 x ]
+  even<even : ∀ {x y} → x < y → 2[1+ x ] < 2[1+ y ]
+  even<odd  : ∀ {x y} → x < y → 2[1+ x ] < 1+[2 y ]
+  odd<even  : ∀ {x y} → x < y ⊎ x ≡ y → 1+[2 x ] < 2[1+ y ]
+  odd<odd   : ∀ {x y} → x < y → 1+[2 x ] < 1+[2 y ]
+
+  -- In these constructors "even" stands for nonzero even.
+{-
+Example. Explanation for  even<odd x<y :
+
+2(1+x) < 1+2y  ~  1+2(1+x) ≤ 1+2y  ~  2(1+x) ≤ 2y  ~  1+x ≤ y  ~ x < y
+
+For  odd<even (inj₁ x<y) :    1+2x < 2(1+y)  ~  2(1+x) ≤ 2(1+y)  ~  x ≤ y
+-}
+
+_>_ :  Rel Bin 0ℓ
+_>_ =  flip _<_
+
+_≤_ :  Rel Bin 0ℓ
+x ≤ y =  x < y ⊎ x ≡ y
+
+_≥_ :  Rel Bin 0ℓ
+_≥_ =  flip _≤_
+
+_≮_ :  Rel Bin 0ℓ
+_≮_ x =  ¬_ ∘ _<_ x
+
+_≯_ :  Rel Bin 0ℓ
+_≯_ x =  ¬_ ∘ _>_ x
+
+_≰_ :  Rel Bin 0ℓ
+_≰_ x =  ¬_ ∘ _≤_ x
+
+_≱_ :  Rel Bin 0ℓ
+_≱_ x =  ¬_ ∘ _≥_ x
+
+≮0 : ∀ {x} → x ≮ zero    -- to use instead of λ()
+≮0 ()
+
+1+[2x]<2[1+x] :  ∀ x → 1+[2 x ] < 2[1+ x ]
+1+[2x]<2[1+x] x =  odd<even (inj₂ refl)
+
+<-irrefl : Irreflexive _≡_ _<_
+<-irrefl {zero}     {zero}      _  ()
+<-irrefl {zero}     {2[1+ _ ]}  ()
+<-irrefl {zero}     {1+[2 _ ]}  ()
+<-irrefl {2[1+ _ ]} {zero}      ()
+<-irrefl {2[1+ x ]} {2[1+ .x ]} refl (even<even x<x) =  <-irrefl refl x<x
+<-irrefl {1+[2 x ]} {1+[2 .x ]} refl (odd<odd x<x)   =  <-irrefl refl x<x
+
+<⇒≢ :  _<_ ⇒ _≢_
+<⇒≢ {x} {_} x<y x≡y =  <-irrefl {x} {x} refl x<x
```

Comment:
You can simplify this proof a lot by pattern matching on `x≡y`. This goes for many of the other proofs below as well. It would be worth trying it everywhere you use a propositional equality.

Reply:
Replaced with ``<⇒≢ x<x refl =  <-irrefl refl x<x``

Pattern matching on `` x≡y`` is applied in several places
(though the proofs become less explicit and sometimes more difficult to understand
for a reader).